File: frontend/src/app/(site)/tournaments/[id]/_components/registration/validation.ts
import type {
  BuiltInFieldConfig,
  CustomFieldDefinition,
  FieldValidationConfig,
} from "@/types/registration.types";

const BUILT_IN_LABELS: Record<string, string> = {
  battle_tag: "BattleTag",
  smurf_tags: "Smurf Accounts",
  discord_nick: "Discord",
  twitch_nick: "Twitch",
  notes: "Notes",
};

const TEXT_VALIDATION_FIELD_TYPES = new Set<CustomFieldDefinition["type"]>(["text", "number", "url"]);
const DEFAULT_BATTLE_TAG_REGEX = String.raw`([\w0-9]{2,12}#[0-9]{4,})`;
const DEFAULT_BATTLE_TAG_VALIDATION: FieldValidationConfig = {
  regex: DEFAULT_BATTLE_TAG_REGEX,
};

function normalizeBattleTag(value: string): string {
  return value.trim().replace(/\s*#\s*/g, "#").replace(/\s+/g, "");
}

export function normalizeBuiltInFieldValue(fieldKey: string, value: string): string {
  const trimmed = value.trim();
  if (fieldKey === "battle_tag" || fieldKey === "smurf_tags") {
    return normalizeBattleTag(trimmed);
  }
  return trimmed;
}

function buildRegex(validation?: FieldValidationConfig | null): RegExp | null {
  const pattern = validation?.regex?.trim();
  if (!pattern) {
    return null;
  }

  try {
    return new RegExp(`^(?:${pattern})$`);
  } catch {
    return null;
  }
}

function resolveBuiltInValidation(
  fieldKey: string,
  validation?: FieldValidationConfig | null,
): FieldValidationConfig | null {
  if (validation?.regex?.trim()) {
    return validation;
  }
  if (fieldKey === "battle_tag") {
    return {
      ...DEFAULT_BATTLE_TAG_VALIDATION,
      ...(validation?.error_message?.trim() ? { error_message: validation.error_message } : {}),
    };
  }
  return validation ?? null;
}

function getErrorMessage(
  validation: FieldValidationConfig | null | undefined,
  label: string,
): string {
  return validation?.error_message?.trim() || `${label} format is invalid.`;
}

export function getBuiltInFieldValidationError(
  fieldKey: string,
  value: string,
  config?: BuiltInFieldConfig,
): string | null {
  return getBuiltInValueValidationError(fieldKey, value, config);
}

export function getBuiltInValueValidationError(
  fieldKey: string,
  value: string,
  config?: BuiltInFieldConfig,
): string | null {
  const validation = resolveBuiltInValidation(fieldKey, config?.validation);
  const regex = buildRegex(validation);
  const candidate = normalizeBuiltInFieldValue(fieldKey, value);
  if (!regex || !candidate) {
    return null;
  }
  if (regex.test(candidate)) {
    return null;
  }

  return getErrorMessage(validation, BUILT_IN_LABELS[fieldKey] ?? fieldKey);
}

export function getBuiltInListValidationError(
  fieldKey: string,
  values: string[],
  config?: BuiltInFieldConfig,
): string | null {
  if (values.length === 0) {
    return null;
  }

  for (const value of values) {
    const error = getBuiltInValueValidationError(fieldKey, value, config);
    if (error) {
      return error;
    }
  }

  return null;
}

export function supportsCustomFieldValidation(
  field: Pick<CustomFieldDefinition, "type">,
): boolean {
  return TEXT_VALIDATION_FIELD_TYPES.has(field.type);
}

export function getCustomFieldValidationError(
  field: CustomFieldDefinition,
  value: string,
): string | null {
  if (!supportsCustomFieldValidation(field)) {
    return null;
  }

  const regex = buildRegex(field.validation);
  const rawValue = value.trim();
  if (!regex || !rawValue) {
    return null;
  }

  if (regex.test(rawValue)) {
    return null;
  }

  return getErrorMessage(field.validation, field.label);
}

export function getFirstLiveValidationError(
  fieldErrors: Record<string, string | null | undefined>,
  fieldKeys: string[],
): string | null {
  for (const fieldKey of fieldKeys) {
    const error = fieldErrors[fieldKey];
    if (error) {
      return error;
    }
  }

  return null;
}

export function getStepDisplayValidationError(
  liveValidationError: string | null,
  stepValidationError: string | null,
): string | null {
  if (liveValidationError) {
    return null;
  }

  return stepValidationError;
}

File: backend/app-service/src/services/registration/validation.py
"""Validation helpers for public registration submissions."""

from __future__ import annotations

import re
from typing import Any

from fastapi import HTTPException, status

from src.schemas.registration import (
    BuiltInFieldConfig,
    CustomFieldDefinition,
    RegistrationCreate,
    RegistrationUpdate,
)

BATTLE_TAG_FIELDS = {"battle_tag", "smurf_tags"}
TEXTUAL_CUSTOM_FIELD_TYPES = {"text", "number", "url"}


def _canonicalize_battle_tag(value: str | None) -> str:
    text = (value or "").strip()
    text = re.sub(r"\s*#\s*", "#", text)
    return text.replace(" ", "").strip()


def _coerce_built_in_field_config(value: Any) -> BuiltInFieldConfig:
    if isinstance(value, BuiltInFieldConfig):
        return value
    return BuiltInFieldConfig.model_validate(value or {})


def _coerce_custom_field_definition(value: Any) -> CustomFieldDefinition:
    if isinstance(value, CustomFieldDefinition):
        return value
    return CustomFieldDefinition.model_validate(value or {})


def _compile_fullmatch_pattern(pattern: str | None) -> re.Pattern[str] | None:
    if pattern is None:
        return None
    normalized = pattern.strip()
    if not normalized:
        return None
    return re.compile(normalized)


def _validation_error(
    message: str,
) -> None:
    raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_CONTENT, detail=message)


def _matches_pattern(pattern: re.Pattern[str] | None, value: str) -> bool:
    if pattern is None:
        return True
    return pattern.fullmatch(value) is not None


def _validate_required_text(
    *,
    value: str | None,
    label: str,
) -> None:
    if (value or "").strip():
        return
    _validation_error(f"{label} is required.")


def _validate_required_list(
    *,
    values: list[str] | None,
    label: str,
) -> None:
    if values:
        return
    _validation_error(f"{label} is required.")


def _validate_string_pattern(
    *,
    value: str | None,
    config: BuiltInFieldConfig | CustomFieldDefinition,
    label: str,
    normalize_battle_tag: bool = False,
) -> None:
    raw_value = (value or "").strip()
    if not raw_value:
        return

    validation = config.validation
    pattern = _compile_fullmatch_pattern(validation.regex if validation else None)
    if pattern is None:
        return

    candidate = _canonicalize_battle_tag(raw_value) if normalize_battle_tag else raw_value
    if _matches_pattern(pattern, candidate):
        return

    _validation_error(validation.error_message or f"{label} format is invalid.")


def _validate_list_pattern(
    *,
    values: list[str] | None,
    config: BuiltInFieldConfig,
    label: str,
) -> None:
    if not values:
        return

    validation = config.validation
    pattern = _compile_fullmatch_pattern(validation.regex if validation else None)
    if pattern is None:
        return

    for value in values:
        candidate = _canonicalize_battle_tag(value)
        if candidate and _matches_pattern(pattern, candidate):
            continue
        _validation_error(validation.error_message or f"{label} format is invalid.")


def _validate_checkbox_requirement(
    *,
    value: Any,
    field: CustomFieldDefinition,
) -> None:
    if value is not None and value != "":
        return
    _validation_error(f"Fill in the required field: {field.label}.")


def _validate_custom_field(
    *,
    field: CustomFieldDefinition,
    value: Any,
    provided: bool,
    partial: bool,
) -> None:
    if partial and not provided:
        return

    if field.required:
        if field.type == "checkbox":
            _validate_checkbox_requirement(value=value, field=field)
        elif not str(value or "").strip():
            _validation_error(f"Fill in the required field: {field.label}.")

    if field.type not in TEXTUAL_CUSTOM_FIELD_TYPES:
        return

    _validate_string_pattern(
        value=None if value is None else str(value),
        config=field,
        label=field.label,
    )


def validate_registration_input(
    form: Any,
    payload: RegistrationCreate | RegistrationUpdate,
    *,
    partial: bool = False,
) -> None:
    built_in_fields = {
        key: _coerce_built_in_field_config(value)
        for key, value in (getattr(form, "built_in_fields_json", None) or {}).items()
    }
    custom_fields = [
        _coerce_custom_field_definition(value)
        for value in (getattr(form, "custom_fields_json", None) or [])
    ]
    provided_fields = payload.model_fields_set if partial else None

    built_in_payload_values: dict[str, Any] = {
        "battle_tag": getattr(payload, "battle_tag", None),
        "smurf_tags": getattr(payload, "smurf_tags", None),
        "discord_nick": getattr(payload, "discord_nick", None),
        "twitch_nick": getattr(payload, "twitch_nick", None),
        "notes": getattr(payload, "notes", None),
        "stream_pov": getattr(payload, "stream_pov", None),
        "roles": getattr(payload, "roles", None),
    }

    for field_key, label in (
        ("battle_tag", "BattleTag"),
        ("smurf_tags", "Smurf Accounts"),
        ("discord_nick", "Discord"),
        ("twitch_nick", "Twitch"),
        ("notes", "Notes"),
    ):
        config = built_in_fields.get(field_key)
        if config is None or not config.enabled:
            continue
        if partial and provided_fields is not None and field_key not in provided_fields:
            continue

        value = built_in_payload_values[field_key]
        if config.required:
            if field_key == "smurf_tags":
                _validate_required_list(values=value, label=label)
            else:
                _validate_required_text(value=value, label=label)

        if field_key == "smurf_tags":
            _validate_list_pattern(values=value, config=config, label=label)
        else:
            _validate_string_pattern(
                value=value,
                config=config,
                label=label,
                normalize_battle_tag=field_key in BATTLE_TAG_FIELDS,
            )

    primary_role_config = built_in_fields.get("primary_role")
    if primary_role_config and primary_role_config.enabled and primary_role_config.required:
        if not partial or (provided_fields is not None and "roles" in provided_fields):
            roles = built_in_payload_values["roles"] or []
            if not any(getattr(role, "is_primary", False) for role in roles):
                _validation_error("Primary Role is required.")

    additional_roles_config = built_in_fields.get("additional_roles")
    if additional_roles_config and additional_roles_config.enabled and additional_roles_config.required:
        if not partial or (provided_fields is not None and "roles" in provided_fields):
            roles = built_in_payload_values["roles"] or []
            if not any(not getattr(role, "is_primary", False) for role in roles):
                _validation_error("At least one additional role is required.")

    custom_values = getattr(payload, "custom_fields", None) or {}
    if not isinstance(custom_values, dict):
        custom_values = {}

    for field in custom_fields:
        provided = field.key in custom_values
        _validate_custom_field(
            field=field,
            value=custom_values.get(field.key),
            provided=provided,
            partial=partial,
        )
